app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenter.kt

```diff
@@ -148,6 +155,49 @@ class HomePresenter(
         environmentManager.addEnvironmentListener(this::class) { refreshTokens() }
     }
 
+    private fun handleDeeplinks() {
+        launchSupervisor {
+            deeplinksManager
+                .subscribeOnDeeplinks(
+                    setOf(
+                        DeeplinkTarget.BUY,
+                        DeeplinkTarget.SEND,
+                        DeeplinkTarget.SWAP,
+                        DeeplinkTarget.CASH_OUT
+                    )
+                )
+                .collect { data ->
+                    when (data.target) {
+                        DeeplinkTarget.BUY -> {
+                            handleBuyDeeplink(data)
+                        }
+                        DeeplinkTarget.SEND -> {
+                            // fixme hack! waiting for tokens to load, probably it's better to show progress
+                            waitForCondition(1000) { state.tokens.isNotEmpty() }
+                            onSendClicked(SearchOpenedFromScreen.MAIN)
+                        }
+                        DeeplinkTarget.SWAP -> {
+                            if (data.hasArgNotNull("from") && data.hasArgNotNull("to")) {
+                                val amount = data.args["amount"]?.toBigDecimalOrNull()?.toPlainString() ?: "0"
+                                view?.showSwapWithArgs(
+                                    tokenA = data.args["from"] as String,
```

Comment:
to remove this casting you can create a method data.requireArg("from") or invoke data.args.getValue("from")